data/src/main/java/com/linkedin/data/collections/CheckedMap.java

```diff
@@ -299,6 +326,27 @@ private final void checkMutability()
     }
   }
 
+  /**
+   * Add a change listener to be notified of changes to the underlying map.
+   *
+   * <p>This class internally maintains weak references to the listeners to avoid leaking them.</p>
+   *
+   * @param listener The listener to register.
+   */
+  public final void addChangeListener(BiFunction<K, V, Void> listener)
+  {
+    if (_changeListeners == null)
+    {
+      // Change listeners are mostly used by map wrappers, and most maps will only have 1
+      // wrapper, so initialize list with a capacity of 1.
+      _changeListeners = new ArrayList<>(1);
+    }
+
+    // Maintain a weak reference to to the listener to avoid leaking the wrapper beyond its
+    // lifetime.
+    _changeListeners.add(new WeakReference<>(listener));
```

Comment:
Just curious, how/and why do we want to put WeakReference here.

Reply:
As the comment says we don't want to leak the wrapper record/union beyond its lifetime. Without a weakreference the record/union will live on as long as the datamap lives on.